### PR TITLE
feat(cluster.tf): add support to set initial release channel version

### LINKED ISSUE #1460: Module doesnt honor specified GKE master version
### TL;DR

When both the release channel and master version are specified the module ignores the master version which creates a discrepancy in the intended cluster version vs what the module is creating. 

https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/private-cluster/cluster.tf#L57

### Expected behavior

If you user specifies the version module should either use that version or fail provisioning with error message version not supported

### Observed behavior

We use this module to deploy GKE clusters. We use a Stable release channel but control the master version so that we can reliably deploy the same K8s version across our fleet. 

### Terraform Configuration

```hcl
module "gke" {
  source             = "terraform-google-modules/kubernetes-engine/google//modules/private-cluster"
  version            = "23.3.0"
  project_id         = var.AccountID
  name               = var.ClusterName
  kubernetes_version = local.latest_gke_patched_version
  regional           = true
  region             = var.Region
  network            = var.ClusterName
  subnetwork         = var.ServiceSubnets[0]
  ip_range_pods      = var.PodRange
  ip_range_services  = var.ClusterServiceRangeName
  horizontal_pod_autoscaling = true
  database_encryption        = local.cluster_encryption_config
  enable_private_endpoint  = false
  enable_private_nodes     = true
  remove_default_node_pool = true
  network_policy           = true
  monitoring_service       = "none"
  master_ipv4_cidr_block   = "10.0.64.0/28"
  logging_service = "logging.googleapis.com/kubernetes"
  cluster_resource_labels           = local.tags
  node_pools                 = local.node_pools
  master_authorized_networks = local.master_authorized_networks
  
  # Release Channel configuration
  release_channel = var.release_channel_version
  maintenance_exclusions = [{
    name = "minor_upgrade_exclusion"
    start_time = timestamp()
    end_time = timeadd(timestamp(),local.end_time_in_hrs)
    exclusion_scope = "NO_MINOR_UPGRADES" # Scope should be one of: NO_UPGRADES | NO_MINOR_UPGRADES | NO_MINOR_OR_NODE_UPGRADES
  }]
}
```

### Terraform Version

```sh
1.0.11
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by slimatic**
This change adds support for specifying an initial version to use when a release channel is being used. The new logic sets the `min_master_version` variable to the value of the locally defined `master_version` when `release_channel` is `null` or `"UNSPECIFIED"`. When a release channel is specified, `min_master_version` is set to the value of `kubernetes_version` if it is defined, and falls back to the value of `master_version` otherwise.

Fixes #1460

### CODE REVIEW

File: cluster.tf [hunk 1]
```diff
@@ -69,7 +69,7 @@ resource "google_container_cluster" "primary" {
     disabled = var.disable_default_snat
   }
 
-  min_master_version = var.release_channel == null || var.release_channel == "UNSPECIFIED" ? local.master_version : null
+  min_master_version = var.release_channel == null || var.release_channel == "UNSPECIFIED" ? local.master_version : var.kubernetes_version == null ? null : var.kubernetes_version
```
Comment (apeabody) [inline cluster.tf:72]: Was this intended to be`null`?

```suggestion
  min_master_version = var.release_channel == null || var.release_channel == "UNSPECIFIED" ? local.master_version : var.kubernetes_version == "latest" ? null : var.kubernetes_version
```

**Review by apeabody — CHANGES_REQUESTED** [review]
Thanks for the contribution @slimatic!

If we don't have one already, please include one of the test fixtures setting `kubernetes_version = "latest"` so this new functionally is covered by the Integration tests.

https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/test/fixtures

**Comment by slimatic** [comment]
> /gcbrun

@apeabody fixed the tests and addressed the lint failures.

**Comment by slimatic** [comment]
> /gcbrun

@apeabody applied `make build` changes to README.md

**Comment by slimatic** [comment]
@apeabody I noticed the cicd build is failing. Is my attention required on the failure? Please advise.

**Comment by apeabody** [comment]
> @apeabody I noticed the cicd build is failing. Is my attention required on the failure? Please advise.

I think it's unrelated, hopefully resolved by https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1637

**Comment by slimatic** [comment]
> Thanks for the contribution @slimatic!
> 
> If we don't have one already, please include one of the test fixtures setting `kubernetes_version = "latest"` so this new functionally is covered by the Integration tests.
> 
> https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/test/fixtures

@apeabody the requested change has been added.

**Comment by slimatic** [comment]
> > Thanks for the contribution @slimatic!
> > If we don't have one already, please include one of the test fixtures setting `kubernetes_version = "latest"` so this new functionally is covered by the Integration tests.
> > https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/test/fixtures
> 
> @apeabody the requested change has been added.

@apeabody apologies, I accidentally closed the PR when I was commenting.

**Comment by slimatic** [comment]
@apeabody just triggered a new merge of the master branch.
could you also advise on what is needed to successfully merge the pull request?

**Comment by apeabody** [comment]
> @apeabody just triggered a new merge of the master branch. could you also advise on what is needed to successfully merge the pull request?

Thanks @slimatic, I'll re-run the CI.  Per merge timeline, what are the potential impacts to existing clusters?  Just need to determine if this is OK for immediate minor release now, or if it will need to be a major release with migration documentation.

**Comment by slimatic** [comment]
> > @apeabody just triggered a new merge of the master branch. could you also advise on what is needed to successfully merge the pull request?
> 
> Thanks @slimatic, I'll re-run the CI. Per merge timeline, what are the potential impacts to existing clusters? Just need to determine if this is OK for immediate minor release now, or if it will need to be a major release with migration documentation.

@apeabody Thanks for providing some insight on the merge timelines.

**Comment by slimatic** [comment]
> 



> > @apeabody just triggered a new merge of the master branch. could you also advise on what is needed to successfully merge the pull request?
> 
> Thanks @slimatic, I'll re-run the CI. Per merge timeline, what are the potential impacts to existing clusters? Just need to determine if this is OK for immediate minor release now, or if it will need to be a major release with migration documentation.


@apeabody were the following questions were directed to me? 
- "Per merge timeline, what are the potential impacts to existing clusters? "
-  "OK for immediate minor release now, or if it will need to be a major release "

**Comment by apeabody** [comment]
Hi @slimatic - Thanks, yes, do you have any insight on the potential impacts to existing clusters?

**Comment by slimatic** [comment]
> Hi @slimatic - Thanks, yes, do you have any insight on the potential impacts to existing clusters?

Yes, based on my tests, making this change will have minimal impact on existing clusters. If the Kubernetes version is modified and applied to an existing cluster, Terraform will upgrade the cluster only if the specified version is greater than the current version. If the version is smaller than the existing version, the terraform apply command will complete successfully without upgrading the cluster version.

**Review by bharathkkb — APPROVED** [review]
Thanks @slimatic and @apeabody. I did a quick test to validate and observed similar behavior as noted by @slimatic so it looks okay to release as minor.

**Comment by apeabody** [comment]
> Thanks @slimatic and @apeabody. I did a quick test to validate and observed similar behavior as noted by @slimatic so it looks okay to release as minor.

Thanks @bharathkkb - That was above and beyond!

**Review by apeabody — APPROVED** [review]

**Comment by apeabody** [comment]
Will merge once the CI is all green.

**Pull request merged**